### ISSUE TITLE
BUG: large number of fails with macOS 15.1 using Accelerate

### ISSUE THREAD

**Issue by andyfaff** [scipy/scipy#21862]
### Describe your issue.

I'm seeing a large number of fails on macOS 15.1 (Sequoia, ARM M1) using Accelerate. I don't see those fails with scipy-openblas32. This leads me to believe that there's a problem with Accelerate, or how we're using it.

I did this test in a new conda environment, did `git clean -xdf`, etc.

The test log is too long to insert here, you can find it in a gist [here](https://gist.github.com/andyfaff/b964e2ecbfe73ca5dcfbfe14adaed89a)

@tylerjereddy, do you have any issues with Sequoia?

### Reproducing Code Example

```python
# conda create -c conda-forge -n dev312 python=3.12
# conda activate dev312
# pip install numpy cython ninja meson pythran pybind11 doit click rich_click pydevtool

python dev.py build --with-accelerate
python dev.py -n test -s linalg
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
numpy=

Project version: 1.15.0.dev0+git20241110.b919b4a
C compiler for the host machine: cc (clang 16.0.0 "Apple clang version 16.0.0 (clang-1600.0.26.4)")
C linker for the host machine: cc ld64 classic  PROJECT:ld64
C++ compiler for the host machine: c++ (clang 16.0.0 "Apple clang version 16.0.0 (clang-1600.0.26.4)")
C++ linker for the host machine: c++ ld64 classic  PROJECT:ld64
Cython compiler for the host machine: cython (cython 3.0.11)
Host machine cpu family: aarch64
Host machine cpu: aarch64
Program python3 found: YES (/Users/andrew/miniforge3/envs/dev312/bin/python3.12)
Found pkg-config: YES (/usr/local/bin/pkg-config) 0.29.2
Run-time dependency python found: YES 3.12
Program cython found: YES (/Users/andrew/miniforge3/envs/dev312/bin/cython)
Compiler for C supports arguments -Wno-unused-but-set-variable: YES 
Compiler for C supports arguments -Wno-unused-function: YES 
Compiler for C supports arguments -Wno-conversion: YES 
Compiler for C supports arguments -Wno-misleading-indentation: YES 
Library m found: YES
Fortran compiler for the host machine: gfortran (gcc 13.2.0 "GNU Fortran (Homebrew GCC 13.2.0) 13.2.0")
Fortran linker for the host machine: gfortran ld64 classic  PROJECT:ld64
Compiler for Fortran supports arguments -Wno-conversion: YES 
Compiler for C supports link arguments -Wl,-ld_classic: YES 
Compiler for C supports link arguments -Wl,-dead_strip: YES 
Checking if "-Wl,--version-script" : links: NO 
Program tools/generate_f2pymod.py found: YES (/Users/andrew/miniforge3/envs/dev312/bin/python3.12 /Users/andrew/Documents/Andy/programming/scipy/tools/generate_f2pymod.py)
Program scipy/_build_utils/tempita.py found: YES (/Users/andrew/miniforge3/envs/dev312/bin/python3.12 /Users/andrew/Documents/Andy/programming/scipy/scipy/_build_utils/tempita.py)
Program pythran found: YES 0.17.0 0.17.0 (/Users/andrew/miniforge3/envs/dev312/bin/pythran)
Found CMake: /usr/local/bin/cmake (3.24.0)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency xsimd found: NO (tried pkgconfig, framework and cmake)
Run-time dependency threads found: YES
Library npymath found: YES
pybind11-config found: YES (/Users/andrew/miniforge3/envs/dev312/bin/pybind11-config) 2.13.6
Run-time dependency pybind11 found: YES 2.13.6
Program f2py found: YES (/Users/andrew/miniforge3/envs/dev312/bin/f2py)
Run-time dependency accelerate found: YES (/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Accelerate.framework)


(dev312) 192-168-1-105:scipy andrew$ pip list
Package            Version
------------------ -----------
attrs              24.2.0
beniget            0.4.2.post1
certifi            2024.8.30
charset-normalizer 3.4.0
click              8.1.7
cloudpickle        3.1.0
Cython             3.0.11
doit               0.36.0
gast               0.6.0
hypothesis         6.118.7
idna               3.10
importlib_metadata 8.5.0
iniconfig          2.0.0
markdown-it-py     3.0.0
mdurl              0.1.2
meson              1.5.2
ninja              1.11.1.1
numpy              2.1.3
packaging          24.2
pip                24.3.1
platformdirs       4.3.6
pluggy             1.5.0
ply                3.11
pooch              1.8.2
pybind11           2.13.6
pydevtool          0.3.0
Pygments           2.18.0
pytest             8.3.3
pythran            0.17.0
requests           2.32.3
rich               13.9.4
rich-click         1.8.3
scipy-openblas32   0.3.28.0.2
setuptools         75.3.0
sortedcontainers   2.4.0
typing_extensions  4.12.2
urllib3            2.2.3
wheel              0.45.0
zipp               3.21.0
```

**Comment by tylerjereddy** [comment]
> @tylerjereddy, do you have any issues with Sequoia?

I'm still on macOS `14.7` so don't have much to add yet. I believe we may have some in-house policies on when we're allowed to upgrade or something like that.

Of course, I have no problems with accelerate at `14.7`.

**Comment by andyfaff** [comment]
The test fails are visible in #21863, https://github.com/scipy/scipy/actions/runs/11770116674/job/32782100147?pr=21863.

This is going to be important to sort out before the next release.

**Comment by tylerjereddy** [comment]
Understood, I added the milestone.

**Comment by andyfaff** [comment]
@rgommers, I can't remember the handle of the engineer working for Apple who we speak to about Accelerate. Is it worth notifying him of this issue?

**Comment by tylerjereddy** [comment]
you might be referring to @Developer-Ecosystem-Engineering

**Comment by rgommers** [comment]
Hmm, that doesn't look great. @andyfaff I suspect that the `macosx_14_0` wheels for SciPy 1.14.1 will show the same failures when run on macOS 15 (at least if it's an Accelerate problem), could you confirm that?

There are a ton of failures in the linked gist, but many are testing the same functions. These are the failures for unique functionality:
```
FAILED scipy/linalg/tests/test_decomp.py::TestEig::test_make_eigvals - Assert...
FAILED scipy/linalg/tests/test_decomp.py::TestOrdQZ::* (many tests)
FAILED scipy/linalg/tests/test_decomp.py::TestOrdQZWorkspaceSize::test_decompose
FAILED scipy/linalg/tests/test_decomp_update.py::TestQRdelete_D::* (many tests)
FAILED scipy/linalg/tests/test_decomp_update.py::TestQRinsert_D::* (many tests)
FAILED scipy/linalg/tests/test_decomp_update.py::TestQRupdate_D::* (many tests)
FAILED scipy/linalg/tests/test_lapack.py::test_gees_trexc[complex128] - Asser...
FAILED scipy/linalg/tests/test_lapack.py::test_gges_tgexc[complex128] - Asser...
FAILED scipy/linalg/tests/test_lapack.py::test_gees_trsen[complex128] - Asser...
FAILED scipy/linalg/tests/test_lapack.py::test_gges_tgsen[complex128] - Asser...
FAILED scipy/linalg/tests/test_solvers.py::TestSolveDiscreteAre::test_solve_discrete_are[0-case0]
FAILED scipy/linalg/tests/test_solvers.py::TestSolveDiscreteAre::test_solve_discrete_are[1-case1]
FAILED scipy/linalg/tests/test_solvers.py::TestSolveDiscreteAre::test_solve_discrete_are[2-case2]
```
It looks like accuracy loss in a LAPACK function for a `complex double` eigenvalue calculation, or something else in that neighborhood.

`solve_discrete_are` calls `ordqz`, so those have the same root cause. `ordqz` in turn uses the `gges` and `tgsen` LAPACK functions. So probably the most relevant failures here are:
```
test_gees_trexc[complex128]
test_gges_tgexc[complex128] 
test_gees_trsen[complex128]
test_gges_tgsen[complex128]
```

**Comment by ilayn** [comment]
Might be that accelerate is on an older lapack too. Not sure it is related but things like https://github.com/Reference-LAPACK/lapack/issues/475 might make a difference

**Comment by rgommers** [comment]
> Might be that accelerate is on an older lapack too

It was actually upgraded in macOS 15.0, from LAPACK 3.9.1 to 3.11.0. It may be a bug in newer LAPACK.

**Comment by andyfaff** [comment]
@Developer-ecosystem-engineering are you able to help us out based on this information? What is the process for raising this with the team responsible for Accelerate?

**Comment by Developer-Ecosystem-Engineering** [comment]
Hi! 

Log a [Feedback](https://developer.apple.com/bug-reporting/) and feel free to @ us here. At a glance, seems like this is all `complex128`.  Taking a deeper look, give us a few days thanks!

**Comment by andyfaff** [comment]
@Developer-Ecosystem-Engineering I have submitted feedback via FB15782411. I've included the steps required to reproduce the problems, but not in the form of individually compilable programs. Rather, it gives the steps required to visualise the problems via running of the scipy test suite.

**Comment by Developer-Ecosystem-Engineering** [comment]
macOS 15.2 beta 4 [release notes](https://developer.apple.com/documentation/macos-release-notes/macos-15_2-release-notes) highlight this issue 

> Resolved an issue impacting the accuracy of some linear algebra operations (*lartg). These functions have been restored to their 3.9.1 implementation. (139760965) (FB15782411)

Please let us know if there are any other issues @andyfaff , et al. Thank you again for the report.

**Comment by andyfaff** [comment]
@Developer-Ecosystem-Engineering thank you for quick response here, it's much appreciated.

**Comment by andyfaff** [comment]
@rgommers, @tylerjereddy 

This issue will exist for macOS 15.0 <= version < 15.2. 15.2 will contain the upstream fix. Will it be necessary to add marks for known test fails?

**Comment by andyfaff** [comment]
@Developer-Ecosystem-Engineering PR #21945 skips several tests that were failing (a test is done to see if `15.0 <= version < 15.2`), but there are still fails that I've not skipped yet.

I think before I start skipping all the rest of the tests that are failing I'd like to get confirmation that the bug fixes to Accelerate that will be shipped with macOS 15.2 are going to solve the problem. There are a couple of possibilities:

1. Can you build scipy with `python dev.py build --with-accelerate` and run the test suite (`python dev.py test`).
2. -or- is there someway of me getting access to the fixed Accelerate functionality so I can test it? (pre-release, etc)

**Comment by tylerjereddy** [comment]
> Will it be necessary to add marks for known test fails?

Not sure, but I suppose the importance of skipping may depend somewhat on the release schedule for OS `15.2`. If the average pip install user doesn't run our testsuite they might silently end up with issues either way I suppose.

**Comment by andyfaff** [comment]
I've tried using a virtual machine (15.1.0) with the latest Xcode beta (16.2-beta-3) which is supposed to target 15.2. However, this didn't pass the test suite. I couldn't find a way to create a VM with 15.2 in it.

**Comment by rgommers** [comment]
Okay, I upgraded to 15.1.1, reproduced the failures, then took a deep breath and installed the 15.2 beta on my work machine - and confirmed it fixes all the test failures. 🎉 

So we should be all set here. macOS 15.2 should come out sometime in December (very likely at least, Apple never gives hard dates beforehand AFAIK), and we're releasing SciPy 1.15.0 in early January. Hence I would not go and add 100+ skips for a problem that is already resolved. We can leave the issue open for visibility and close it as soon as macOS 15.2 is available in my opinion.

* * *

The one follow-up action I'd like to take is to build vanilla Netlib BLAS/LAPACK 3.11.0 and 3.12.0 from source and see if we get the same failures there. If we do, then there is a problem with accuracy in LAPACK that needs discussing. If not, the problem was something specific to the 3.9.0 -> 3.11.0 upgrade within Accelerate. My guess is that it's the latter, since OpenBLAS has already included 3.11.0 and 3.12.0, but we need to be sure.

**Comment by andyfaff** [comment]
I've just updated to the recently released 15.2 update, and it fixes the test fails as well. Thank you @Developer-Ecosystem-Engineering

**Comment by Developer-Ecosystem-Engineering** [comment]
Thank you all for the report and diligent investigation!
